gluten-core/src/main/scala/org/apache/gluten/extension/columnar/OffloadSingleNode.scala

```diff
@@ -181,7 +185,127 @@ case class OffloadJoin() extends OffloadSingleNode with LogLevelUtil {
+      var newOutput = output
+      for ((_, newAttr) <- replacedExprs) {
+        if (!newOutput.exists(attr => attr.exprId == newAttr.exprId)) {
+          newOutput = newOutput :+ newAttr
+        }
+      }
+      newOutput
+    }
+    def genNewProjectList(projectList: Seq[NamedExpression]): Seq[NamedExpression] = {
+      var newProjectList = projectList
+      for ((_, newAttr) <- replacedExprs) {
+        if (!newProjectList.exists(attr => attr.exprId == newAttr.exprId)) {
+          newProjectList = newProjectList :+ newAttr.toAttribute
+        }
+      }
+      newProjectList
+    }
+    plan match {
+      case f: FileSourceScanExec =>
+        f.copy(output = genNewOutput(f.output))
+      case f: FileSourceScanExecTransformer =>
+        f.copy(output = genNewOutput(f.output))
+      case b: BatchScanExec =>
+        b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+      case b: BatchScanExecTransformer =>
+        b.copy(output = genNewOutput(b.output).asInstanceOf[Seq[AttributeReference]])
+      case p @ ProjectExec(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case p @ ProjectExecTransformer(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case _ => plan.withNewChildren(plan.children.map(addMetadataCol(_, replacedExprs)))
+    }
+  }
+
+  private def genProjectExec(projectExec: ProjectExec): SparkPlan = {
+    def findScanNodes(plan: SparkPlan): Seq[SparkPlan] = {
+      plan match {
+        case f @ (_: FileSourceScanExec | _: AbstractFileSourceScanExec |
+            _: DataSourceV2ScanExecBase) =>
+          Seq(f)
+        case o => o.children.flatMap(findScanNodes)
+      }
+    }
+
+    if (!TransformHints.isNotTransformable(projectExec)) {
+      logWarning(s"Columnar Processing for ${projectExec.getClass} is currently supported.")
+      ProjectExecTransformer(projectExec.projectList, projectExec.child)
+    } else if (BackendsApiManager.getSettings.supportNativeInputFileNameExpr()) {
+      val addHint = AddTransformHintRule()
+      val newProjectList = projectExec.projectList.filterNot(isInputFileNameExpr)
+      val newProjectExec = ProjectExec(newProjectList, projectExec.child)
+      addHint.apply(newProjectExec)
+      if (TransformHints.isNotTransformable(newProjectExec)) {
+        // Project is still not transformable after remove `input_file_name` expressions.
+        projectExec
+      } else {
+        // the project with `input_file_name` expression should have at most
+        // one data source, reference:
+        // https://github.com/apache/spark/blob/e459674127e7b21e2767cc62d10ea6f1f941936c
+        // /sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala#L506
+        val scanChild = findScanNodes(projectExec)
+        if (scanChild.isEmpty || TransformHints.isNotTransformable(scanChild(0))) {
```

Comment:
> `the project with `input_file_name` expression should have at most one data source`

If it's true, we'd add a assertion to ensure it. E.g.,

```
assert(scanChild.size == 1)
```

Also the name of `scanChild` is misleading. Please change to `leafScans`.